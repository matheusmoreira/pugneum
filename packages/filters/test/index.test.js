'use strict';

const fs = require('fs');
const path = require('path');
const assert = require('assert');

const lex = require('pugneum-lexer');
const parse = require('pugneum-parser');
const filter = require('../');

const filename = path.basename(__filename);

var customFilters = {
  custom: function(str, options) {
    return 'BEGIN' + str + 'END';
  },
  'custom-with-options': function(str, options) {
    expect(options.option).toBe('value');
    expect(options.number).toBe('2'); // no automatic parsing of option values
    return 'BEGIN OPTIONS' + str + 'END OPTIONS';
  }
};

test('filters can be used', () => {
  const source = `
p
  :custom
    Filters can be used.
`;

  const ast = parse(lex(source, {filename}), {
    filename,
    src: source,
  });

  const output = filter(ast, customFilters);
  expect(output).toMatchSnapshot();
});

process.chdir(__dirname + '/../');

var testCases;

testCases = fs.readdirSync(__dirname + '/cases').filter(function(name) {
  return /\.input\.json$/.test(name);
});

testCases.forEach(function(filename) {
  function read(path) {
    return fs.readFileSync(__dirname + '/cases/' + path, 'utf8');
  }

  test('cases/' + filename, function() {
    var actualAst = JSON.stringify(
      filter(JSON.parse(read(filename)), customFilters),
      null,
      '  '
    );
    expect(actualAst).toMatchSnapshot();
  });
});

testCases.forEach(function(filename) {
  function read(path) {
    return fs.readFileSync(__dirname + '/errors/' + path, 'utf8');
  }

  test('errors/' + filename, function() {
    var actual;
    try {
      filter(JSON.parse(read(filename)), customFilters);
      throw new Error('Expected ' + filename + ' to throw an exception.');
    } catch (ex) {
      if (!ex || !ex.code || ex.code.indexOf('PUGNEUM:') !== 0) throw ex;
      actual = {
        msg: ex.msg,
        code: ex.code,
        line: ex.line,
      };
    }
    expect(actual).toMatchSnapshot();
  });
});
