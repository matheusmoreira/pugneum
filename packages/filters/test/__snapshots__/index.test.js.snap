// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters can be used 1`] = `
{
  "filename": "index.test.js",
  "line": 0,
  "nodes": [
    {
      "attributeBlocks": [],
      "attrs": [],
      "block": {
        "filename": "index.test.js",
        "line": 2,
        "nodes": [
          {
            "attrs": [],
            "block": {
              "filename": "index.test.js",
              "line": 3,
              "nodes": [
                {
                  "column": 5,
                  "filename": "index.test.js",
                  "line": 4,
                  "type": "Text",
                  "val": "Filters can be used.",
                },
              ],
              "type": "Block",
            },
            "column": 3,
            "filename": "index.test.js",
            "line": 3,
            "name": "custom",
            "type": "Text",
            "val": "BEGINFilters can be used.END",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "filename": "index.test.js",
      "isInline": false,
      "line": 2,
      "name": "p",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;

exports[`filters can be used with options 1`] = `
{
  "filename": "index.test.js",
  "line": 0,
  "nodes": [
    {
      "attributeBlocks": [],
      "attrs": [],
      "block": {
        "filename": "index.test.js",
        "line": 2,
        "nodes": [
          {
            "attrs": [
              {
                "column": 24,
                "filename": "index.test.js",
                "line": 3,
                "mustEscape": true,
                "name": "option",
                "val": "value",
              },
              {
                "column": 36,
                "filename": "index.test.js",
                "line": 3,
                "mustEscape": true,
                "name": "number",
                "val": "2",
              },
            ],
            "block": {
              "filename": "index.test.js",
              "line": 3,
              "nodes": [
                {
                  "column": 5,
                  "filename": "index.test.js",
                  "line": 4,
                  "type": "Text",
                  "val": "Filters can be used with options.",
                },
                {
                  "column": 1,
                  "filename": "index.test.js",
                  "line": 5,
                  "type": "Text",
                  "val": "
",
                },
                {
                  "column": 5,
                  "filename": "index.test.js",
                  "line": 5,
                  "type": "Text",
                  "val": "The values aren't parsed though.",
                },
                {
                  "column": 1,
                  "filename": "index.test.js",
                  "line": 6,
                  "type": "Text",
                  "val": "
",
                },
                {
                  "column": 5,
                  "filename": "index.test.js",
                  "line": 6,
                  "type": "Text",
                  "val": "They're just strings.",
                },
              ],
              "type": "Block",
            },
            "column": 3,
            "filename": "index.test.js",
            "line": 3,
            "name": "custom-with-options",
            "type": "Text",
            "val": "BEGIN OPTIONSFilters can be used with options.
The values aren't parsed though.
They're just strings.END OPTIONS",
          },
        ],
        "type": "Block",
      },
      "column": 1,
      "filename": "index.test.js",
      "isInline": false,
      "line": 2,
      "name": "p",
      "selfClosing": false,
      "type": "Tag",
    },
  ],
  "type": "Block",
}
`;
